waterfall/golang/aoa/aoa.go

```diff
@@ -0,0 +1,279 @@
+	serial       = "00000012345678"
+	uri          = "http://play.google.com/store/apps/details?id=com.google.waterfall.usb"
+
+	aoaClass    = 255
+	aoaSubclass = 255
+	aoaProtocol = 0
+)
+
+var (
+	// Product ID to use when in accessory mode + ADB
+	usbAccessoryADBProductID int = 0x2D01
+
+	// Product ids with accessory mode enabled.
+	usbAccessoryIds = map[int]bool{
+		0x2D00: true,
+		0x2D01: true,
+		0x2D04: true,
+		0x2D05: true,
+	}
+
+	// vendor specific requests
+	vendorInXfer  uint8 = 0xC0
+	vendorOutXfer uint8 = 0x40
+
+	// requests
+	getProtocolRequest    uint8 = 51
+	sendStringRequest     uint8 = 52
+	startAccessoryRequest uint8 = 53
+
+	manufacturerIDX uint16 = 0
+	modelIDX        uint16 = 1
+	descriptionIDX  uint16 = 2
+	versionIDX      uint16 = 3
+	uriIDX          uint16 = 4
+	serialIDX       uint16 = 5
+)
+
+type ReadWriteCloser struct {
+	dev    *gousb.Device
+	config *gousb.Config
+	intf   *gousb.Interface
+	in     *gousb.InEndpoint
+	out    *gousb.OutEndpoint
+}
+
+func (u *ReadWriteCloser) Close() error {
+	u.intf.Close() // intf.Close does not return an error
+
+	if err := u.config.Close(); err != nil {
+		return err
+	}
+	return u.dev.Close()
+}
+
+func (u *ReadWriteCloser) Read(in []byte) (int, error) {
+	return u.in.Read(in)
+}
+
+func (u *ReadWriteCloser) Write(out []byte) (int, error) {
+	return u.out.Write(out)
+}
+
+func sendString(dev *gousb.Device, idx uint16, str string) error {
+	_, err := dev.Control(vendorOutXfer, sendStringRequest, 0, idx, []byte(str))
+	return err
+}
+
+// switchToAccessory switched the USB device to Accessory mode following the protocol
+// described in https://source.android.com/devices/accessories/aoa2.
+func switchToAccessory(dev *gousb.Device) error {
+	if err := sendString(dev, manufacturerIDX, manufacturer); err != nil {
+		return err
+	}
+	if err := sendString(dev, modelIDX, model); err != nil {
+		return err
+	}
+	if err := sendString(dev, descriptionIDX, description); err != nil {
+		return err
+	}
+	if err := sendString(dev, versionIDX, version); err != nil {
+		return err
+	}
+	if err := sendString(dev, uriIDX, uri); err != nil {
+		return err
+	}
+	if err := sendString(dev, serialIDX, serial); err != nil {
+		return err
+	}
+	if _, err := dev.Control(vendorOutXfer, startAccessoryRequest, 0, 0, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+func makeReadWriteCloser(dev *gousb.Device, config *gousb.Config, i *gousb.Interface) (*ReadWriteCloser, error) {
+	rw := &ReadWriteCloser{
+		dev:    dev,
+		config: config,
+		intf:   i,
+	}
+	for _, desc := range i.Setting.Endpoints {
+		switch desc.Direction {
+		case gousb.EndpointDirectionIn:
+			in, err := i.InEndpoint(desc.Number)
+			if err != nil {
+				return nil, err
+			}
+			rw.in = in
+		case gousb.EndpointDirectionOut:
+			out, err := i.OutEndpoint(desc.Number)
+			if err != nil {
+				return nil, err
+			}
+			rw.out = out
+		}
+	}
+	return rw, nil
+}
+
+func getActiveConfig(dev *gousb.Device) (*gousb.Config, error) {
```

Comment:
Drop get prefix here and below

Reply:
done

Reply:
not resolving since im not sure if the comments disappear